compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSuspendOnInvalidTypeChecker.kt

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.declaration
+
+import org.jetbrains.kotlin.fir.FirElement
+import org.jetbrains.kotlin.fir.FirFakeSourceElementKind
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.declarations.FirFile
+import org.jetbrains.kotlin.fir.types.FirFunctionTypeRef
+import org.jetbrains.kotlin.fir.types.FirResolvedTypeRef
+import org.jetbrains.kotlin.fir.types.FirTypeRefWithNullability
+import org.jetbrains.kotlin.fir.visitors.FirDefaultVisitorVoid
+import org.jetbrains.kotlin.lexer.KtTokens
+
+object FirSuspendOnInvalidTypeChecker : FirFileChecker() {
+    override fun check(declaration: FirFile, context: CheckerContext, reporter: DiagnosticReporter) {
+        // Traverse the entire file to look for source type refs (i.e., Fir(Dynamic|User|Function)TypeRef).
+        declaration.accept(object : FirDefaultVisitorVoid() {
+            override fun visitElement(element: FirElement) {
+                val source = element.source ?: return
+                if (source.kind is FirFakeSourceElementKind) return
+                element.acceptChildren(this)
+            }
+
+            override fun visitResolvedTypeRef(resolvedTypeRef: FirResolvedTypeRef) {
+                // We will only get to the source type refs by visiting delegatedTypeRef.
+                super.visitResolvedTypeRef(resolvedTypeRef)
+                resolvedTypeRef.delegatedTypeRef?.accept(this)
+            }
+
+            override fun visitTypeRefWithNullability(typeRefWithNullability: FirTypeRefWithNullability) {
+                // `suspend` is invalid for non-function types (i.e., FirDynamicTypeRef or FirUserTypeRef).
+                //
+                // It is also invalid for nullable function types, e.g., `suspend (() -> Int)?`.
+                // The correct way to denote a nullable suspend function type is `(suspend () -> Int)?`.
+                // In both cases, the source for the FirFunctionTypeRef is the TYPE_REFERENCE element.
+                // In the invalid case, the modifier list is in the source TYPE_REFERENCE element.
+                // But in the correct case, the modifier list is in the child NULLABLE_TYPE element.
+                // Therefore, if the FirFunctionTypeRef is marked nullable, it is invalid to HAVE the `suspend` modifier
+                // on the source element, even though it seems counter-intuitive.
+                val suspendModifier = typeRefWithNullability.getModifier(KtTokens.SUSPEND_KEYWORD) ?: return
```

Comment:
Just to make sure.

>  it is invalid to HAVE the `suspend` modifier on the source element

Is this done by `typeRef...getModifier(...) ?: return` ? If so, this looks a little bit implicit..., but I don't have a better idea. :\

Reply:
No, the point of that line (`typeRef...getModifier(...) ?: return`) is that if there is no `suspend` modifier, there's nothing to check and nothing to report on.

Where "if the FirFunctionTypeRef is marked nullable, it is invalid to HAVE the `suspend` modifier" is checked is in the 2nd OR clause in `if (typeRefWithNullability !is FirFunctionTypeRef || typeRefWithNullability.isMarkedNullable)`. It's equivalent to `if (typeRefWithNullability !is FirFunctionTypeRef || (typeRefWithNullability  is FirFunctionTypeRef && typeRefWithNullability.isMarkedNullable))`.

`!A || (A && B) == !A || B` (remember we chatted about this offline 🙂 )

Reply:
Yeah, I do remember that, and it is also read like implication: `!A || B == A -> !B`. That is, if `typeRef...` is functional type, it shouldn't be `isMarkedNullable`. So, I understood "if the FirFunctionTypeRef is marked nullable" matched that 2nd OR clause.

I was literally talking about "it is invalid to HAVE the suspend modifier on the source element". When I read code comments, I expected like:
- checking non-function types, since it's invalid
- otherwise, checking isMarkedNullable, since nullable function types are invalid
- now, "it is invalid to HAVE the suspend modifier on the source element"
- uh, wait, having `suspend` has been checked already? hm, isn't it somehow implicit?

But, anyway, it's not a big deal. But maybe separating/adding code comments might be helpful, e.g.,
```kt
// This checker is literally checking `suspend` modifier on invalid types
val suspendModifier = ... ?: return

// `suspend` is invalid for non-function types...
// It is also invalid for nullable function types...
if (typeRef... !is FirFunctionTypeRef || typeRef...isMarkedNullable) {
  ...
}
```

Reply:
Got it. So I'm just going to move the retrieval of the modifier to before the comment block. Didn't add a comment for the retrieval statement since I think it's obvious enough that we're looking for `suspend`. I also de-capitalized "HAVE" (I think the emphasis adds to the confusion instead of clarity.)